cmd/xl-sets.go

```diff
@@ -0,0 +1,1101 @@
+	}()
+
+	hashReader, err := hash.NewReader(pipeReader, objInfo.Size, "", "")
+	if err != nil {
+		pipeReader.CloseWithError(err)
+		return objInfo, toObjectErr(errors.Trace(err), destBucket, destObject)
+	}
+
+	objInfo, err = destSet.PutObject(destBucket, destObject, hashReader, metadata)
+	if err != nil {
+		pipeReader.CloseWithError(err)
+		return objInfo, err
+	}
+
+	// Explicitly close the reader.
+	pipeReader.Close()
+
+	return objInfo, nil
+}
+
+// Returns function "listDir" of the type listDirFunc.
+// isLeaf - is used by listDir function to check if an entry is a leaf or non-leaf entry.
+// disks - used for doing disk.ListDir(). Sets passes set of disks.
+func listDirSetsFactory(isLeaf isLeafFunc, treeWalkIgnoredErrs []error, sets ...[]StorageAPI) listDirFunc {
+	listDirInternal := func(bucket, prefixDir, prefixEntry string, disks []StorageAPI) (entries []string, err error) {
+		for _, disk := range disks {
+			if disk == nil {
+				continue
+			}
+
+			entries, err = disk.ListDir(bucket, prefixDir)
+			if err != nil {
+				// For any reason disk was deleted or goes offline, continue
+				// and list from other disks if possible.
+				if errors.IsErrIgnored(err, treeWalkIgnoredErrs...) {
+					continue
+				}
+				return nil, errors.Trace(err)
+			}
+
+			break
+		}
+		return entries, nil
+	}
+
+	// listDir - lists all the entries at a given prefix and given entry in the prefix.
+	listDir := func(bucket, prefixDir, prefixEntry string) (mergedEntries []string, delayIsLeaf bool, err error) {
+		for _, disks := range sets {
+			var entries []string
+			entries, err = listDirInternal(bucket, prefixDir, prefixEntry, disks)
+			if err != nil {
+				return nil, false, err
+			}
+
+			var newEntries []string
+			// Find elements in entries which are not in mergedEntries
+			for _, entry := range entries {
+				idx := sort.SearchStrings(mergedEntries, entry)
+				// if entry is already present in mergedEntries don't add.
+				if idx < len(mergedEntries) && mergedEntries[idx] == entry {
+					continue
+				}
+				newEntries = append(newEntries, entry)
+			}
+
+			if len(newEntries) > 0 {
+				// Merge the entries and sort it.
+				mergedEntries = append(mergedEntries, newEntries...)
+				sort.Strings(mergedEntries)
+			}
+		}
+		mergedEntries, delayIsLeaf = filterListEntries(bucket, prefixDir, mergedEntries, prefixEntry, isLeaf)
+		return mergedEntries, delayIsLeaf, nil
+	}
+	return listDir
+}
+
+// ListObjects - implements listing of objects across sets, each set is independently
+// listed and subsequently merge lexically sorted inside listDirSetsFactory(). Resulting
+// value through the walk channel receives the data properly lexically sorted.
+func (s *xlSets) ListObjects(bucket, prefix, marker, delimiter string, maxKeys int) (result ListObjectsInfo, err error) {
+	// validate all the inputs for listObjects
+	if err = checkListObjsArgs(bucket, prefix, marker, delimiter, s); err != nil {
+		return result, err
+	}
```

Comment:
```
        if err := checkListObjsArgs(bucket, prefix, marker, delimiter, xl); err != nil {
                return loi, err
        }

        // With max keys of zero we have reached eof, return right here.
        if maxKeys == 0 {
                return loi, nil
        }

        // For delimiter and prefix as '/' we do not list anything at all
        // since according to s3 spec we stop at the 'delimiter' along
        // with the prefix. On a flat namespace with 'prefix' as '/'
        // we don't have any entries, since all the keys are of form 'keyName/...'
        if delimiter == slashSeparator && prefix == slashSeparator {
                return loi, nil
        }

        // Over flowing count - reset to maxObjectList.
        if maxKeys < 0 || maxKeys > maxObjectList {
                maxKeys = maxObjectList
        }

```

these checks are done in xl-v1-list-objects.go

Reply:
BTW, xl-v1-list-objects.go can be removed right?

Reply:
It cannot be just yet. @krishnasrinivas there are lot of test cases which depend on it.